File: plugins/harness/skills/harness-install/assets/uv/runtime/rules/scaffold_leaks.py
#!/usr/bin/env -S uv run
# -*- coding: utf-8 -*-
"""
Forbid scaffold leaks rule.
"""
from collections.abc import Iterable

import re
from pathlib import Path

from harness_check_rule import Finding, HarnessCheckRule

from .utils import is_relative_to, read_text, relative, safe_file_or_walk, severity_for


def strip_markdown_code(text: str) -> str:
    """
    Remove Markdown code blocks and inline code spans before prose-level checks.
    """
    stripped_lines: list[str] = []
    in_fence = False
    fence_marker = ""
    for line in text.splitlines():
        fence_match = re.match(r" {0,3}(`{3,}|~{3,})", line)
        if fence_match:
            marker = fence_match.group(1)[0]
            if not in_fence:
                in_fence = True
                fence_marker = marker
            elif marker == fence_marker:
                in_fence = False
            stripped_lines.append("")
            continue
        if in_fence:
            stripped_lines.append("")
            continue
        stripped_lines.append(re.sub(r"`+[^`\n]*`+", "", line))
    return "\n".join(stripped_lines)


class ScaffoldLeaksRule(HarnessCheckRule):
    """
    Validate scaffoldLeaks check.
    """
    category = "scaffoldLeaks"

    def applies(self, manifest: dict) -> bool:
        """
        Check if this rule applies to the manifest.
        """
        section = manifest.get(self.category)
        if not isinstance(section, dict):
            return False
        return section.get("enabled", True) is not False

    def validate(self, root: Path, manifest: dict) -> Iterable[Finding]:
        """
        Validate scaffoldLeaks check.
        """
        section = manifest.get(self.category, {})
        if not isinstance(section, dict):
            return []
        params = section.get("parameters", {})
        if not isinstance(params, dict):
            return []
        scope = params.get("scope", {})
        if not isinstance(scope, dict):
            return []
        bases_data = scope.get("bases", [])
        excluded_data = scope.get("excludedSubtrees", [])
        exts_data = scope.get("extensions", [])
        if not isinstance(bases_data, list):
            return []
        resolved_root = root.resolve()
        active_roots = tuple(
            candidate for b in bases_data if isinstance(b, str)
            for candidate in (root / b,)
            if is_relative_to(candidate.resolve(), resolved_root)
        )
        excluded_paths = tuple(
            candidate for e in (excluded_data if isinstance(excluded_data, list) else ())
            if isinstance(e, str)
            for candidate in (root / e,)
            if is_relative_to(candidate.resolve(), resolved_root)
        )
        extensions = frozenset(
            f".{ext}" for ext in (exts_data if isinstance(exts_data, list) else [])
            if isinstance(ext, str)
        ) or frozenset({".md", ".txt"})
        patterns_data = params.get("patterns", [])
        if not isinstance(patterns_data, list):
            return []
        compiled_patterns = tuple(
            (re.compile(item.get("pattern")), item.get("label"))
            for item in patterns_data
            if isinstance(item, dict)
            and isinstance(item.get("pattern"), str)
            and isinstance(item.get("label"), str)
            and (lambda p=item.get("pattern"): True if re.compile(p) else False)()
        )
        messages = section.get("messages", {})
        if not isinstance(messages, dict):
            return []
        return [
            Finding(
                severity_for(manifest, self.category),
                self.category,
                messages.get("default", "{label} in active asset: {file}").format(
                    label=label, file=relative(path)
                ),
            )
            for base in active_roots
            for path in safe_file_or_walk(base)
            if path.is_file()
            and is_relative_to(path.resolve(), resolved_root)
            and path.suffix in extensions
            and not any(path == ex or ex in path.parents for ex in excluded_paths)
            for text in (strip_markdown_code(read_text(path)),)
            for pattern, label in compiled_patterns
            if pattern.search(text)
        ]


RULE: HarnessCheckRule = ScaffoldLeaksRule()

File: plugins/harness/skills/harness-install/assets/bun/runtime/rules/text/scaffold-leaks.ts
#!/usr/bin/env bun
// -*- coding: utf-8 -*-
import type {
  Finding,
  HarnessCheckRule,
  RuleContext,
} from "../harness-check-rule";

/**
 * Remove Markdown code blocks and inline code spans before prose-level checks.
 */
const stripMarkdownCode = (text: string): string => {
  const lines: string[] = [];
  let inFence = false;
  let fenceMarker = "";
  for (const line of text.split(/\r?\n/)) {
    const fenceMatch = /^( {0,3})(`{3,}|~{3,})/.exec(line);
    if (fenceMatch) {
      const marker = fenceMatch[2]?.charAt(0) ?? "";
      if (!inFence) {
        inFence = true;
        fenceMarker = marker;
        lines.push("");
      } else if (marker === fenceMarker) {
        inFence = false;
        lines.push("");
      }
    } else if (inFence) {
      lines.push("");
    } else {
      lines.push(line.replace(/`+[^`\n]*`+/g, ""));
    }
  }
  return lines.join("\n");
};

/**
 * Compile a pattern that may carry a Python-style inline flag prefix (e.g. `(?m)`)
 * into a JavaScript RegExp by lifting the flags into the constructor argument.
 */
const compilePortableRegex = (pattern: string): RegExp => {
  const inlineFlagMatch = /^\(\?([a-z]+)\)/.exec(pattern);
  if (inlineFlagMatch?.[1]) {
    return new RegExp(
      pattern.slice(inlineFlagMatch[0].length),
      inlineFlagMatch[1],
    );
  }
  return new RegExp(pattern);
};

/**
 * Scaffold leak regex paired with its reporting label.
 */
interface ScaffoldPattern {
  readonly regex: RegExp;
  readonly label: string;
}

/**
 * Forbid scaffold/placeholder patterns in active assets.
 */
export const scaffoldLeaksRule: HarnessCheckRule = {
  category: "scaffoldLeaks",
  applies(ctx: RuleContext): boolean {
    const section = ctx.manifest.raw.scaffoldLeaks;
    if (typeof section !== "object" || section === null) {
      return false;
    }
    const enabled = (section as { enabled?: unknown }).enabled;
    return (
      enabled !== false &&
      ctx.readStringArray(
        ctx.readJsonObject(
          ctx.readJsonObject((section as Record<string, unknown>).parameters)
            .scope,
        ).bases,
      ).length > 0
    );
  },

  validate(ctx: RuleContext): readonly Finding[] {
    const parameters = ctx.readJsonObject(
      ctx.readJsonObject(ctx.manifest.raw.scaffoldLeaks).parameters,
    );
    const scope = ctx.readJsonObject(parameters.scope);
    const bases = ctx.readStringArray(scope.bases);
    const excludedSubtrees = ctx.readStringArray(scope.excludedSubtrees);
    const extensions = ctx.readStringArray(scope.extensions);
    const patterns: readonly ScaffoldPattern[] = Array.isArray(
      parameters.patterns,
    )
      ? (parameters.patterns as unknown[])
          .filter(
            (item): item is Record<string, unknown> =>
              typeof item === "object" && item !== null,
          )
          .map((obj) => ({
            patternStr: typeof obj.pattern === "string" ? obj.pattern : "",
            labelStr: typeof obj.label === "string" ? obj.label : "",
          }))
          .filter(
            ({ patternStr, labelStr }) => patternStr !== "" && labelStr !== "",
          )
          .map(
            ({ patternStr, labelStr }) =>
              ({ regex: compilePortableRegex(patternStr), label: labelStr }),
          )
      : [];
    return bases.flatMap((base) => {
      const [files, warnings] = ctx.collectFilesUnder(base);
      return warnings.concat(
        files
          .filter((file) => {
            const isExcluded = excludedSubtrees.some(
              (subtree) => file === subtree || file.startsWith(`${subtree}/`),
            );
            const ext = /\.([a-z0-9]+)$/.exec(file)?.[1] ?? "";
            return !isExcluded && extensions.includes(ext);
          })
          .flatMap((file) => {
            const text = ctx.read(file);
            return patterns
              .filter((patternEntry) => patternEntry.regex.test(stripMarkdownCode(text)))
              .flatMap((patternEntry) => {
                const match = patternEntry.regex.exec(stripMarkdownCode(text));
                const positions = match
                  ? (() => {
                      const before = text.slice(0, match.index);
                      const startLinVal = before.split("\n").length;
                      const startColVal = match.index - (before.lastIndexOf("\n") + 1) + 1;
                      const matchLines = match[0].split("\n");
                      const endLinVal = startLinVal + matchLines.length - 1;
                      const endColVal = matchLines.length === 1
                        ? startColVal + match[0].length
                        : (matchLines[matchLines.length - 1]?.length ?? 0) + 1;
                      return {
                        startLine: startLinVal,
                        startColumn: startColVal,
                        endLine: endLinVal,
                        endColumn: endColVal,
                      };
                    })()
                  : { startLine: 1, startColumn: 1, endLine: 1, endColumn: 1 };
                return [
                  {
                    severity: ctx.severityOf("scaffoldLeaks"),
                    category: "scaffoldLeaks",
                    message: `${patternEntry.label} in active asset: ${file}`,
                    file,
                    ...positions,
                    fix: {
                      description: `replace scaffold pattern: ${patternEntry.label}`,
                      safety: "manual",
                    },
                  },
                ];
              });
          }),
      );
    });
  },
};
